### ISSUE TITLE
bosh agent rewrites partition mistakenly during VM reboot

### ISSUE THREAD

**Issue by mattcui** [cloudfoundry/bosh-agent#92]
We found bosh agent always rewrites the partition table every time when the VM is rebooted even if the disk is already partitioned and formatted. In the part of code of determining if needing to do partition, there is a problem where to get the device size at the line of  [`size, err := p.GetDeviceSizeInBytes(partitionPath)`](https://github.com/cloudfoundry/bosh-agent/blob/master/platform/disk/sfdisk_partitioner.go#L181)

There will return the following error when running this line:

```
2016-08-31_10:01:11.18516 [Cmd Runner] 2016/08/31 10:01:11 DEBUG - Running command 'sfdisk -s /dev/mapper/3600a09803830357a392448666a38304f1'
2016-08-31_10:01:11.18813 [Cmd Runner] 2016/08/31 10:01:11 DEBUG - Stdout:
2016-08-31_10:01:11.18814 [Cmd Runner] 2016/08/31 10:01:11 DEBUG - Stderr: /dev/mapper/3600a09803830357a392448666a38304f1: No such file or directory
2016-08-31_10:01:11.18815
2016-08-31_10:01:11.18815 sfdisk: cannot open /dev/mapper/3600a09803830357a392448666a38304f1 for reading
```

It results in the problem where the variable `size` is not assigned, and it is still 0 as the initial value. The simplest way to fix the issue would be to modify `partitionPath` to `devicePath` at the line of [`size, err := p.GetDeviceSizeInBytes(partitionPath)`](https://github.com/cloudfoundry/bosh-agent/blob/master/platform/disk/sfdisk_partitioner.go#L181), but I am not sure if it's a good way to fix this issue. Need BOSH guys to give some comments/advice.

This issue has been already under discussion with @cppforlife, @dpb587-pivotal and @maximilien, I raised it here for better tracking.

**Comment by cppforlife** [comment]
Added a story: https://www.pivotaltracker.com/story/show/129727181 to properly handle the error; however, this wouldnt fix the "sfdisk: cannot open" error. Are you sure changes for multipath disk handling are properly accounting for partitions?

**Comment by mattcui** [comment]
@cppforlife Yes, it's due to multipath, I submitted [PR 93](https://github.com/cloudfoundry/bosh-agent/pull/93), please help review. Thank you!

/cc @maximilien @dpb587-pivotal

**Comment by mattcui** [comment]
Agent log after making this change:

```
2016-09-05_12:13:05.45238 [linuxPlatform] 2016/09/05 12:13:05 DEBUG - Persistent disk size to be partitioned is: 21474836480, and error is: <nil>
2016-09-05_12:13:05.45240 [linuxPlatform] 2016/09/05 12:13:05 DEBUG - fdisk partitioner was chosen
2016-09-05_12:13:05.45240 [MountPersistentDisk] partitions passed into Partition()
2016-09-05_12:13:05.45242 [[Type: linux, SizeInBytes: 0]]
2016-09-05_12:13:05.45250 [Cmd Runner] 2016/09/05 12:13:05 DEBUG - Running command 'sfdisk -d /dev/mapper/3600a098038303453692447376330554d'
2016-09-05_12:13:05.45551 [Cmd Runner] 2016/09/05 12:13:05 DEBUG - Stdout: # partition table of /dev/mapper/3600a098038303453692447376330554d
2016-09-05_12:13:05.45552 unit: sectors
2016-09-05_12:13:05.45553
2016-09-05_12:13:05.45553 /dev/mapper/3600a098038303453692447376330554d1 : start=        1, size= 41929649, Id=83
2016-09-05_12:13:05.45554 /dev/mapper/3600a098038303453692447376330554d2 : start=        0, size=        0, Id= 0
2016-09-05_12:13:05.45554 /dev/mapper/3600a098038303453692447376330554d3 : start=        0, size=        0, Id= 0
2016-09-05_12:13:05.45555 /dev/mapper/3600a098038303453692447376330554d4 : start=        0, size=        0, Id= 0
2016-09-05_12:13:05.45555 [Cmd Runner] 2016/09/05 12:13:05 DEBUG - Stderr:
2016-09-05_12:13:05.45557 [Cmd Runner] 2016/09/05 12:13:05 DEBUG - Successful: true (0)
2016-09-05_12:13:05.45557 [getPartitions] sfdisk -d result split line: /dev/mapper/3600a098038303453692447376330554d1 : start=        1, size= 41929649, Id=83
2016-09-05_12:13:05.45562 [Cmd Runner] 2016/09/05 12:13:05 DEBUG - Running command 'sfdisk -s /dev/mapper/3600a098038303453692447376330554d-part1'
2016-09-05_12:13:05.46009 [Cmd Runner] 2016/09/05 12:13:05 DEBUG - Stdout: 20964824
2016-09-05_12:13:05.46010 [Cmd Runner] 2016/09/05 12:13:05 DEBUG - Stderr:
2016-09-05_12:13:05.46011 [Cmd Runner] 2016/09/05 12:13:05 DEBUG - Successful: true (0)
2016-09-05_12:13:05.46012 [getPartitions] partition.SizeInBytes is 21467979776
2016-09-05_12:13:05.46013 [getPartitions] sfdisk -d result split line: /dev/mapper/3600a098038303453692447376330554d2 : start=        0, size=        0, Id= 0
2016-09-05_12:13:05.46015 [getPartitions] sfdisk -d result split line: /dev/mapper/3600a098038303453692447376330554d3 : start=        0, size=        0, Id= 0
2016-09-05_12:13:05.46015 [getPartitions] sfdisk -d result split line: /dev/mapper/3600a098038303453692447376330554d4 : start=        0, size=        0, Id= 0
2016-09-05_12:13:05.46021 [Cmd Runner] 2016/09/05 12:13:05 DEBUG - Running command 'sfdisk -s /dev/mapper/3600a098038303453692447376330554d'
2016-09-05_12:13:05.46319 [Cmd Runner] 2016/09/05 12:13:05 DEBUG - Stdout: 20971520
2016-09-05_12:13:05.46320 [Cmd Runner] 2016/09/05 12:13:05 DEBUG - Stderr:
2016-09-05_12:13:05.46321 [Cmd Runner] 2016/09/05 12:13:05 DEBUG - Successful: true (0)
2016-09-05_12:13:05.46324 [SfdiskPartitioner] 2016/09/05 12:13:05 INFO - /dev/mapper/3600a098038303453692447376330554d already partitioned as expected, skipping
```

**Comment by cppforlife** [comment]
@mattcui I believe this issue was resolved after fixing multipath disk lookup? Can be closed?

**Comment by mattcui** [comment]
@cppforlife It's addressed by my [PR 93](https://github.com/cloudfoundry/bosh-agent/issues/93), the commit has been already merged into master branch 1.5 months ago, but it still hasn't been picked by BOSH release code, so we are still waiting for the new BOSH release which includes this fix.  This fix is very important to our CF deployments on Softlayer. Thanks.

/cc @maximilien

**Comment by mattcui** [comment]
I saw the fix has already merged in 3263.8 stemcell, closing this issue. Thanks.
